app/controllers/sessions_controller.rb

```diff
@@ -9,7 +9,43 @@ def create
       session[:user_id] = user.id
       redirect_to root_path, notice: "Welcome back, #{current_user.first_name}"
     else
-      redirect_to login_path
+      process_form_login
     end
   end
+
+  def destroy
+    @current_user = nil
+    session.delete :user_id
+    redirect_to root_path, notice: "Logged out Successfully!"
+  end
+
+  private
+
+  def process_form_login
+    user = RegularUser.find_by_email(params[:email])
+    return if user_is_nil?(user)
+    if user.authenticate(params[:password])
+      login_successful user
+    else
+      password_invalid
+    end
+  end
+
+  def user_is_nil?(user)
+    return if user
+    flash["errors"] = ["Log in failed. Email not valid."]
+    redirect_to login_path
+
+    true
+  end
+
+  def login_successful(user)
+    session[:user_id] = user.id
+    redirect_to root_path, notice: "Welcome back, #{current_user.first_name}"
+  end
+
+  def password_invalid
+    flash["errors"] = ["Password incorrect. Please try again."]
+    redirect_to login_path
+  end
 end
```

Comment:
From my own perspective(I don't really know which is the best practice). I think we are having too many methods involved. The target should be to make code easily readable among developers, and transferring little functionality like `login_successful` could be an extra overhead to the reader. Also, I don't think we need to tell the user specifically the issue with his login details. This could be a drop in the security and conventionally, the login errors are not given with specific reference to either `email` or `password`.